### DISCUSSION TITLE
before_first_request and url_for

### DISCUSSION THREAD

**Discussion by phistrom** [pallets/flask#4978] [Q&A] [answered]
I used to use `before_first_request` to set up a sort of navigation config.
```py
@app.before_first_request
def _setup_navigation(app):
    app.config['navigation'] = [
        {'name': 'Home', 'url': url_for('mysite.home', _external=False)},
        {'name': 'About', 'url': url_for('mysite.about', _external=False)},
        ...
    ]
```
and then I have a `navbar.jinja2` that iterates through this creating links based on the list found in `app.config.navigation` like
```html
<div class="navbar" id="main-nav">
    <ul class="nav navbar-nav">
    {% for nav in config.navigation %}
        <li><a href="{{ nav.url }}">{{ nav.name }}</a></li>
    {% endfor %}
    </ul>
</div>
```

But then `before_first_request` was deprecated. So I am trying to update to use the app factory pattern. But I cannot use `url_for` outside a request context unless I specify a `SERVER_NAME`. But then it's more difficult to do local debugging since I can't just go to `localhost:8000` anymore because that doesn't match the `SERVER_NAME`.

To get around this, I have just added a `before_request` decorator and wrapped it in a test for the existence of my `app.config['navigation']` key.

Perhaps `url_for` could be relaxed to not need a `SERVER_NAME` when `_external=False`?

**Comment by davidism** [comment, marked as answer]
You can set the `SERVER_NAME` to `localhost:8000` during development.

https://flask.palletsprojects.com/en/2.2.x/appcontext/#manually-push-a-context shows how to push an app context during the factory function.

**Reply by phistrom** [reply]
I guess I could push a [test_request_context](https://flask.palletsprojects.com/en/latest/reqcontext/#manually-push-a-context) during app initialization.

**Comment by tachyondecay** [comment]
It’s unclear to me why you aren't just generating these on each request (using ` before_request`) or doing it in the template itself (`url_for` is automatically injected into the template environment for you). This should not create much of a performance impact.

**Reply by phistrom** [reply]
It probably wouldn't create much performance impact to do it every time. In most places I use `url_for` directly in my templates. It just felt easier to make a nice loop in the jinja2 template than to copy and paste the same HTML markup for each link I wanted to add. Regenerating the navigation data structure on every request felt wasteful, and I believe [it was an argument](https://github.com/pallets/flask/issues/4605#:~:text=It%20incurs%20a%20function%20call%20before%20every%20request) for deprecating `before_first_request` in the first place. I suppose I'll just rewrite it to have just the endpoint string (i.e. `'mysite.about'`) and then have `<a href="{{ url_for(nav.url, _external=False) }}">{{ nav.name }}</a>` in the navbar.jinja2 template.

To be clear, I'm fine with the deprecation, but am I wrong for thinking that `url_for` doesn't need a `SERVER_NAME` when generating `_external=False` URLs?
